### DISCUSSION TITLE
/secret-management-with-docker-compose-and-ansible/

### DISCUSSION THREAD

**Discussion by migratecomments** [ironicbadger/blog.ktz.me#6] [Announcements]
Secret Management with docker-compose and Ansible Vault
        
[https://blog.ktz.me/secret-management-with-docker-compose-and-ansible/](https://blog.ktz.me/secret-management-with-docker-compose-and-ansible/)

**Comment by johnnyggalt** [comment]
Thanks for the write-up, Alex. Right now, I see Ansible as overkill for my simple requirements, yet still like the idea of encrypting my at-rest secrets (I commit them in plain text to my self-hosted gitea inside .env files). I wonder if there's a happy middle-ground where something like ansible-vault can be used but without requiring Ansible itself. Docker's Secrets support seems to rely on Docker Swarm, which again seems like too high a bar for me right now, especially as I'm running this on a Synology and I don't even think running Swarm is possible there (could be wrong though).

Do you know of any tool that solely exists to keep at-rest secrets safe but can integrate cleanly with Docker Compose?

**Reply by Jip-Hop** [reply]
Hi @johnnyggalt. Like you I was looking for a middle ground. First [I came up with a solution which uses `ansible-vault` without requiring Ansible itself](https://gist.github.com/Jip-Hop/0b3bb9e2c63300297471c122b44a8d41).

While this worked I found it had several downsides.

- You can't see which environment variables (keys) are present in the vault, unless you decrypt the entire vault
- You need the private key to add new environment variables
- This approach requires wrapping around the `docker compose up` command
- The decrypted environment variable values end up in the container spec and may leak (visible via `docker inspect`, Portainer, logs etc.)

So I tried to migrate my secrets to native file based Docker secrets, but this has 2 issues:
1. You can't safely manage Docker secrets through Git
2. Many images expect secrets as environment variables instead of files

So I decided to develop [ojster](https://github.com/ojster/ojster) to address these issues. It delivers GitOps-safe one-way encrypted secrets, unsealed at container runtime as ephemeral env vars, for Docker Compose.

I hope you find it useful!
